File: AllSamples/Samples/Samples_Source/natural language/langcap/vb/cmdopts.vb
'---------------------------------------------------------------------
'  This file is part of the Microsoft .NET Framework SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------
Imports System
Imports System.Globalization


Public Class CmdOpts
    Public Sub New()
        m_sOutput = ""
        m_Culture = Nothing
        m_fVerbose = False
    End Sub 'New

    Public Property Output() As String
        Get
            Return m_sOutput
        End Get
        Set(ByVal Value As String)
            m_sOutput = Value
        End Set
    End Property

    Public Property Culture() As CultureInfo
        Get
            Return m_Culture
        End Get
        Set(ByVal Value As CultureInfo)
            m_Culture = Value
        End Set
    End Property

    Public Property Verbose() As Boolean
        Get
            Return m_fVerbose
        End Get
        Set(ByVal Value As Boolean)
            m_fVerbose = Value
        End Set
    End Property

    Public Sub DisplayHelp()
        DisplayHelpLine("O", "Output", "The output file (Defaults to Standard Out)")
        DisplayHelpLine("C", "Culture", "Sets the culture of the engine (ex. -C 'en-us' or 'fr-ca').  Defaults to current culture. ")
        DisplayHelpLine("V", "Verbose", "Verbose output")
    End Sub 'DisplayHelp

    ' DisplayHelp
    Public Function ProcessOpts(ByVal args() As String) As Boolean
        Dim i As Integer = 0

        While i < args.Length
            If args(i).StartsWith("-") Then
                Dim thisarg As [String] = args(i)
                thisarg = thisarg.Remove(0, 1).ToLower(System.Globalization.CultureInfo.CurrentCulture)

                If [String].Compare(thisarg, "output") = 0 Or [String].Compare(thisarg, "o") = 0 Then
                    If (i + 1) >= args.Length Then
                        ' make sure there's a next arg to move on to
                        Throw New System.ArgumentException([String].Concat("No Value for ", thisarg))
                    End If
                    ' assign to the next arg
                    m_sOutput = args(i + 1)
                End If

                If [String].Compare(thisarg, "culture") = 0 Or [String].Compare(thisarg, "c") = 0 Then
                    If (i + 1) >= args.Length Then
                        ' make sure there's a next arg to move on to
                        Throw New System.ArgumentException("No Value for ", thisarg)
                    End If
                    ' assign to the next arg
                    m_Culture = getCultureFromArg(args(i + 1))
                End If

                If [String].Compare(thisarg, "verbose") = 0 Or [String].Compare(thisarg, "v") = 0 Then
                    m_fVerbose = True
                End If

                If [String].Compare(thisarg, "help") = 0 Or [String].Compare(thisarg, "?") = 0 Then
                    DisplayHelp()
                    Return False
                End If
            End If
            i = i + 1
        End While ' for
        Return True
    End Function 'ProcessOpts
    ' ProcessOpts

    Private Sub DisplayHelpLine(ByVal shortname As [String], ByVal name As [String], ByVal description As [String])
        Console.WriteLine(" -{0,1} {1,-24} {2}", shortname, name, description)
    End Sub 'DisplayHelpLine


    Private Function getCultureFromArg(ByVal sArg As [String]) As CultureInfo
        Dim culture As CultureInfo = CultureInfo.CurrentCulture

        Try
            culture = New CultureInfo(sArg)
            Return culture
        Catch e As ArgumentException
            Console.Error.WriteLine(e.Message)
            Try
                culture = New CultureInfo(Int32.Parse(sArg, CultureInfo.CurrentCulture.NumberFormat), True)
                Return culture
            Catch ex As ArgumentException
                Console.Error.WriteLine(ex.Message)
                Return CultureInfo.CurrentCulture
            End Try
        End Try
    End Function 'getCultureFromArg

    Private m_sOutput As [String]
    Private m_Culture As CultureInfo
    Private m_fVerbose As Boolean
End Class 'CmdOpts 
' class CmdOpts

File: AllSamples/Samples/Samples_Source/natural language/langcap/csharp/cmdopts.cs
//---------------------------------------------------------------------
//  This file is part of the Microsoft .NET Framework SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------
using System;
using System.Globalization;

namespace Microsoft.Samples.NaturalLanguage
{
	public class CmdOpts
	{
		public CmdOpts()
		{
			m_sOutput = "";
			m_Culture = null;
			m_fVerbose = false;
		}

		public string Output
		{
			get { return m_sOutput; }
			set { m_sOutput = value; }
		}

		public CultureInfo Culture
		{
			get { return m_Culture; }
			set { m_Culture = value; }
		}

		public bool Verbose
		{
			get { return m_fVerbose; }
			set { m_fVerbose = value; }
		}

		public void DisplayHelp()
		{
			DisplayHelpLine("O", "Output", "The output file (Defaults to Standard Out)");
			DisplayHelpLine("C", "Culture", "Sets the culture of the engine (ex. -C 'en-us' or 'fr-ca').  Defaults to current culture. ");
			DisplayHelpLine("V", "Verbose", "Verbose output");
		} // DisplayHelp

		public bool ProcessOpts(string[] args)
		{
			for (int i = 0; i < args.Length; ++i)
			{
				if ((args[i].StartsWith("-")))
				{
					String thisArgument = args[i];

					thisArgument = thisArgument.Remove(0, 1).ToLower(System.Globalization.CultureInfo.CurrentCulture);
					if ((String.Compare(thisArgument, "output") == 0) || (String.Compare(thisArgument, "o") == 0))
					{
						if (++i >= args.Length)
						{   // make sure there's a next arg to move on to
							throw new System.ArgumentException(String.Concat("No Value for ", thisArgument));
						}

						// assign to the next arg
						m_sOutput = args[i];
					}

					if ((String.Compare(thisArgument, "culture") == 0) || (String.Compare(thisArgument, "c") == 0))
					{
						if (++i >= args.Length)
						{   // make sure there's a next arg to move on to
							throw new System.ArgumentException("No Value for ", thisArgument);
						}

						// assign to the next arg
						m_Culture = GetCultureFromArguments(args[i]);
					}

					if ((String.Compare(thisArgument, "verbose") == 0) || (String.Compare(thisArgument, "v") == 0))
					{
						m_fVerbose = true;
					}

					if ((String.Compare(thisArgument, "help") == 0) || (String.Compare(thisArgument, "?") == 0))
					{
						DisplayHelp();
						return false;
					}
				}
			} // for

			return true;
		} // ProcessOpts

		private void DisplayHelpLine(String shortname, String name, String description)
		{
			Console.WriteLine(" -{0,1} {1,-24} {2}", shortname, name, description);
		}

		private CultureInfo GetCultureFromArguments(String arguments)
		{
			CultureInfo culture = CultureInfo.CurrentCulture;

			try
			{
				culture = new CultureInfo(arguments);
				return culture;
			}
			catch (ArgumentException e)
			{
				Console.Error.WriteLine(e.Message);
				try
				{
					culture = new CultureInfo(Int32.Parse(arguments, CultureInfo.CurrentCulture.NumberFormat), true);
					return culture;
				}
				catch (ArgumentException ex)
				{
					Console.Error.WriteLine(ex.Message);
					return CultureInfo.CurrentCulture;
				}
			}
		}

		String m_sOutput;

		CultureInfo m_Culture;

		bool m_fVerbose;
	}; // class CmdOpts
}
